lib/charms/bind/v0/dns_record.py

```diff
@@ -0,0 +1,570 @@
+        """Fetch the DNS domains from the relation."""
+        return self.dns_record_requirer_relation_data.dns_domains
+
+    @property
+    def dns_entries(self) -> Optional[List[RequirerEntry]]:
+        """Fetch the DNS entries from the relation."""
+        return self.dns_record_requirer_relation_data.dns_entries
+
+
+class DNSRecordRequiresEvents(ops.CharmEvents):
+    """DNS record requirer events.
+
+    This class defines the events that a DNS record requirer can emit.
+
+    Attributes:
+        dns_record_request_processed: the DNSRecordRequestProcessed.
+    """
+
+    dns_record_request_processed = ops.EventSource(DNSRecordRequestProcessed)
+
+
+class DNSRecordRequires(ops.Object):
+    """Requirer side of the DNS requires relation.
+
+    Attributes:
+        on: events the provider can emit.
+    """
+
+    on = DNSRecordRequiresEvents()
+
+    def __init__(self, charm: ops.CharmBase, relation_name: str = DEFAULT_RELATION_NAME) -> None:
+        """Construct.
+
+        Args:
+            charm: the provider charm.
+            relation_name: the relation name.
+        """
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+
+    def get_remote_relation_data(self) -> Optional[DNSRecordProviderData]:
+        """Retrieve the remote relation data.
+
+        Returns:
+            DNSRecordProviderData: the relation data.
+        """
+        relation = self.model.get_relation(self.relation_name)
+        return self._get_remote_relation_data(relation) if relation else None
+
+    def _get_remote_relation_data(self, relation: ops.Relation) -> DNSRecordProviderData:
+        """Retrieve the remote relation data.
+
+        Args:
+            relation: the relation to retrieve the data from.
+
+        Returns:
+            DNSRecordProviderData: the relation data.
+        """
+        assert relation.app
+        relation_data = relation.data[relation.app]
+        return DNSRecordProviderData.from_relation_data(relation_data)
+
+    def _is_remote_relation_data_valid(self, relation: ops.Relation) -> bool:
+        """Validate the relation data.
+
+        Args:
+            relation: the relation to validate.
+
+        Returns:
+            true: if the relation data is valid.
+        """
+        try:
+            _ = self._get_remote_relation_data(relation)
+            return True
+        except ValidationError as ex:
+            error_fields = set(
+                itertools.chain.from_iterable(error["loc"] for error in ex.errors())
+            )
+            error_field_str = " ".join(f"{f}" for f in error_fields)
+            logger.warning("Error validation the relation data %s", error_field_str)
+            return False
+
+    def _on_relation_changed(self, event: ops.RelationChangedEvent) -> None:
```

Comment:
Shouldn't we also check the status of the different dns-domains and dns-entries and provide a more meaningful event ? Or emit one event per domain/record ?

Reply:
Yes, will do in a follow up PR